### PR TITLE
build: Bump golang to 1.17.9/Mac M1 Support

### PR DESCRIPTION
**Pull request by Eric-Warehime**
## Summary
Bump go version to 1.17.9 (latest 1.17 version).
It's good to upgrade compiler version regularly, but we also need this for #3808 

Also includes `//+build` => `//go:build` generated via `make sanity`.

Blocking merge on #3909 / #3917 which contains go1.16 version bump.

## Test Plan

Run CI/Perf tests to benchmark against current version.

### CODE REVIEW

File: go.mod
Comment (brianolson) [inline go.mod:17]: lots of little changes to `go.mod`, was `go mod tidy` run?

Comment (Eric-Warehime) [reply go.mod:17]: Yes the build failed without running tidy.

**Comment by Eric-Warehime** [comment]
Additional commit modifies `node.mu` to a `RWLock` and adds `RLock()`/`RUnlock()` calls to fix failures caused by the data race detector found in `node.go`.

File: agreement/msgp_gen_test.go
Comment (cce) [inline agreement/msgp_gen_test.go:1]: Since this file is generated, to avoid having to reapply this change every time these files are generated this would have to be added here too: https://github.com/algorand/msgp/blob/a756cc98119ad08a8fb2aa6ef9a4a844016644b5/printer/print.go#L147-L148

File: node/node.go
Comment (cce) [inline node/node.go:798]: which test did you see this race come from? (and what was the last access to node.ctx recorded by the race detector?)

there are two places where this thread is launched (AlgorandFullNode.Start and SetCatchpointCatchupMode) and I was curious if it was the catchup mode one.

Comment (cce) [reply node/node.go:798]: I found https://app.circleci.com/pipelines/github/algorand/go-algorand/6331/workflows/ff6442e3-7fc2-4654-b109-e8e14a0013d7/jobs/108678 OK I see it is catchpointCatchupTest.exp.

Comment (cce) [reply node/node.go:798]: OK! LGTM.

Comment (cce) [reply node/node.go:798]: That way these threads are bound to be cancelled by the node context that launched them, not whatever happens to be the current value of node.ctx. You wouldn't need the lock this way, either.

File: Makefile
Comment (cce) [inline Makefile:43]: Is this something we should be setting here, or something that an M1 homebrew user should know needs to be in the shell environment already?

Comment (cce) [reply Makefile:43]: I would argue against including this snippet, because strictly speaking homebrew has nothing to do with go-algorand and is not a dependency, and developers using homebrew on M1 will need to have this in their .bash_profile or .zprofile already. (?)

File: node/node.go
Comment (brianolson) [reply node/node.go:798]: All the usage I've seen of Context.Done() is
`case <-ctx.Done():`
... so I'm suspicious of this usage of getting the chan outside of the switch. Maybe it's okay? Maybe I'm not sure `ctx.Done()` always needs to return the same object? cce's point that this should be stable and not need a lock also feels right to me.

File: Makefile
Comment (Eric-Warehime) [reply Makefile:43]: I've taken a look, specifically https://github.com/Homebrew/brew/issues/9177 has a lot of discussion around this, and I'm not sure that I agree.

It's our script which is ultimately installing the boost library to this path via https://github.com/algorand/go-algorand/blob/master/scripts/configure_dev.sh#L80 and the homebrew developers have explicitly said that it's the tools which use brew's software which need to integrate with the changes (things which used to be automatically found by virtue of being in `/usr/` are now in `/opt/homebrew/` on M1).

Homebrew does provide `brew shellenv` to update the user's shell, but this is the output of that:
```
set -gx HOMEBREW_PREFIX "/opt/homebrew";
set -gx HOMEBREW_CELLAR "/opt/homebrew/Cellar";
set -gx HOMEBREW_REPOSITORY "/opt/homebrew";
set -q PATH; or set PATH ''; set -gx PATH "/opt/homebrew/bin" "/opt/homebrew/sbin" $PATH;
set -q MANPATH; or set MANPATH ''; set -gx MANPATH "/opt/homebrew/share/man" $MANPATH;
set -q INFOPATH; or set INFOPATH ''; set -gx INFOPATH "/opt/homebrew/share/info" $INFOPATH;
```
So it seems like it's not expected to be configured in your shell based on that.

Ultimately I'm fine with removing this, but it seems counterintuitive to me that we would `brew install boost` and then expect users to configure library variables themselves before running `make build`. 
Maybe we should be including these environment variable settings in `scripts/configure_dev.sh` instead of the makefile?

File: agreement/msgp_gen_test.go
Comment (Eric-Warehime) [reply agreement/msgp_gen_test.go:1]: ACK, will update.

File: node/node.go
Comment (Eric-Warehime) [reply node/node.go:798]: Let me take another look...Acquiring the lock outside of the for loop seems fine to me. 

Ultimately, from what I understand we _have_ to have the lock because in https://github.com/algorand/go-algorand/blob/master/node/node.go#L1232 we're writing the context variable before the cancel function has been called which means `ctx.Done()` may never be executed since it could retrieve the new channel before the cancel function invokes that branch of the select statement.

Comment (cce) [reply node/node.go:798]: in SetCatchpointCatchupMode, when switching modes, it calls the cancelFunc for the current node.ctx, then it creates a new ctx+cancelFunc pair and sets node.ctx. So unless you have a way to know that the 3 goroutines that are polling on node.ctx.Done() after calling the old cancelFunc, you will probably set off the race detector sometimes..?

The context.Done interface docs say "successive calls to Done return the same value." so I think it is safe to just grab ctx.Done() when each of these 3 goroutines start and keep polling on them.

Comment (algorandskiy) [reply node/node.go:798]: >so I'm suspicious of this usage of getting the chan outside of the switch. Maybe it's okay? Maybe I'm not sure ctx.Done() always needs to return the same object? cce's point that this should be stable and not need a lock also feels right to me.

Looks like Brian and right the [Done()](https://github.com/golang/go/blob/master/src/context/context.go#L358) is way more non-trivial than I expected.

The problem with `SetCatchpointCatchupMode` is it re-assigns `node.ctx` under a lock, and then threads use it without locking. I agree, my proposed fix that Eric implemented does not look correct. The plan was is to go with this fix to unblock others and then re-shuffle `SetCatchpointCatchupMode`  to stop goroutines first, assign, run after - more like `node.Start()` and `node.Stop()`

Comment (Eric-Warehime) [reply node/node.go:798]: > in SetCatchpointCatchupMode, when switching modes, it calls the cancelFunc for the current node.ctx, then it creates a new ctx+cancelFunc pair and sets node.ctx.

Not sure if I'm misunderstanding, but it seems like the exact opposite of that. It [writes a new context](https://github.com/algorand/go-algorand/blob/master/node/node.go#L1232) then [calls the cancel func on the old one](https://github.com/algorand/go-algorand/blob/master/node/node.go#L1235). So if we keep calling `node.ctx.Done()` as we are currently doing, we could end up with the following:

1. Node started, Context<sub>original</sub> initialized, monitoring threads started and polling `<-node.ctx.Done()` receiving on the channel.
2. Catchpoint mode invoked, `node.ctx` overwritten with a new context, Context<sub>new</sub>.
3. Monitoring threads are now receiving the new channel when calling `node.ctx.Done()`
4. Catchpoint mode invokes the cancel func of Context<sub>original</sub>.
5. Monitoring threads are not cancelled because they are monitoring Context<sub>new</sub>.
6. Catchpoint mode finishes and starts a second set of monitoring threads.

> so I think it is safe to just grab ctx.Done() when each of these 3 goroutines start and keep polling on them.
I agree that this probably works as well.

Comment (algorandskiy) [reply node/node.go:798]: >The context.Done interface docs say "successive calls to Done return the same value." 

Right, I looked deeper into `c.done` usage and it is the same, and `Done()` just loads from non-changed `c.done`.

Comment (Eric-Warehime) [reply node/node.go]: I tried something similar to this last night and it was timing out on https://github.com/algorand/go-algorand/blob/master/test/e2e-go/features/catchup/catchpointCatchup_test.go#L224 
Not sure if there are minor differences w/ your changes that will behave differently.

The version I've just pushed tries to address the core problem which is that the node mutex has really high lock contention because it's locking so many different things. So I've split out this into a context-specific mutex. Haven't run into test failures locally with this. Let me know what you think.

Comment (cce) [reply node/node.go]: Hrmm OK well I was hoping if you needed a lock, you would only have to grab the lock once, when the goroutine starts. I think it is a bug to grab the done channel more than once that could lead to orphaned goroutines. Was curious so I wrote up https://github.com/algorand/go-algorand/pull/3955/files to have CI run it...

Comment (Eric-Warehime) [reply node/node.go]: Looks like it's indeed passing. Let's discuss this afternoon pros and cons of each approach and figure out what we want.

Comment (cce) [reply node/node.go]: follow-up — talked with @Eric-Warehime and just explicitly passing the ctx.Done() channel to each goroutine as an argument seemed like the clearest way to avoid worrying about unnecessary synchronization https://github.com/algorand/go-algorand/pull/3955/commits/7878d412188889245a611cf185cc4b4e4ed0274c

File: agreement/msgp_gen_test.go
Comment (Eric-Warehime) [reply agreement/msgp_gen_test.go:1]: https://github.com/algorand/msgp/pull/15

File: Makefile
Comment (winder) [reply Makefile:43]: I agree with Eric. The Makefile needs to ensure the environment is setup appropriately. In this case using `/opt/homebrew` is consistent with our configuration script. The only addition I would make here, is a comment explaining that this change is related to M1 and the new location where homebrew keeps binaries.

File: node/node.go
Comment (winder) [inline node/node.go:410]: @AlgoStephenAkiki fyi - there is going to be at least this merge conflict with your integration branch.

File: scripts/buildtools/go.mod
Comment (winder) [inline scripts/buildtools/go.mod:6]: This is where you'll update msgp after your other PR is merged and tagged

**Review by winder — COMMENTED** [review]
This looks ready to go once your msgp PR is merged and available.

**Review by cce — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

File: scripts/buildtools/go.mod
Comment (algorandskiy) [reply scripts/buildtools/go.mod:6]: @Eric-Warehime please update the version and I'll merge this

Comment (Eric-Warehime) [reply scripts/buildtools/go.mod:6]: Updated. Also had to run `go get` and `go mod tidy` which ended up separating deps into separate `require` blocks. See https://go-review.googlesource.com/c/go/+/325922/ and https://github.com/golang/go/issues/45965 for details.

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**
